wolfssl/wolfcrypt/chacha20_poly1305.h

```diff
@@ -76,6 +101,27 @@ int wc_ChaCha20Poly1305_Decrypt(
                 const byte inAuthTag[CHACHA20_POLY1305_AEAD_AUTHTAG_SIZE],
                 byte* outPlaintext);
 
+WOLFSSL_API
+int wc_ChaCha20Poly1305_CheckTag(
+    const byte authTag[CHACHA20_POLY1305_AEAD_AUTHTAG_SIZE],
+    const byte authTagChk[CHACHA20_POLY1305_AEAD_AUTHTAG_SIZE]);
+
+
+#ifndef NO_CHACHAPOLY_AEAD_IUF
+/* Implementation of AEAD, which includes support for adding
+    data, then final calculation of authentication tag */
+WOLFSSL_API int wc_ChaCha20Poly1305_Init(ChaChaPoly_Aead* aead,
+    const byte inKey[CHACHA20_POLY1305_AEAD_KEYSIZE],
+    const byte inIV[CHACHA20_POLY1305_AEAD_IV_SIZE],
+    int isEncrypt);
+WOLFSSL_API int wc_ChaCha20Poly1305_UpdateAad(ChaChaPoly_Aead* aead,
+    const byte* inAAD, word32 inAADLen);
+WOLFSSL_API int wc_ChaCha20Poly1305_UpdateData(ChaChaPoly_Aead* aead,
+    byte* data, word32 dataLen);
+WOLFSSL_API int wc_ChaCha20Poly1305_Final(ChaChaPoly_Aead* aead,
+    byte outAuthTag[CHACHA20_POLY1305_AEAD_AUTHTAG_SIZE]);
+#endif /* !NO_CHACHAPOLY_AEAD_IUF */
```

Comment:
Why have two update functions? Could you consider consolidating wc_ChaCha20Poly1305_UpdateAad() and wc_ChaCha20Poly1305_UpdateData() into one API simlar to wc_ChaCha20Poly1305_Encrypt() which takes AAD and plaintext?
When using AAD, plaintext will be NULL and vice-versa. 
. Actually, to eliminate duplicate code and save certification cost,  Init/Update/Final should be wrappers of wc_ChaCha20Poly1305_Encrypt/Decrypt and the majority of the changes to support the functionality should go in wc_ChaCha20Poly1305_Encrypt/Decrypt().

Reply:
Having separate AAD and Data is typical for this AEAD and was spec'd by the customer.

Reply:
For the duplicate code I absolutely agree. The existing `wc_ChaCha20Poly1305_Encrypt`/`wc_ChaCha20Poly1305_Decrypt` functions can use the new init/update/final. Also there is a `wc_Poly1305_MAC` function that can be reduced.

I already talked with Todd about this and I'll be doing a separate cleanup to reduce duplicate at a later time. This PR is about adding the feature for the customer only.

Reply:
@dgarske, Thanks, it makes sense why you have two functions. Can you keep the two functions and leverage wc_ChaCha20Poly1305_Encrypt/Decrypt as much as you can to reduce SLOC?

Reply:
For the cert version of this we do not need `wc_ChaCha20Poly1305_Encrypt` and `wc_ChaCha20Poly1305_Decrypt`... they can be removed.

Reply:
@dgarske, excellent!  If the plan is to remove wc_ChaCha20Poly1305_Encrypt/Decrypt from the cert version, it works for me.